aws/amazon-cloudwatch-agent
# Add config option for logging in the AWS SDK.

# Description of the issue
Closes #279
We want to be able to enable logging in the AWS SDK.
We want the log messages to show up in our Agent's log file.


# Description of changes
The `config-translator` program will look for `"aws_sdk_log_level"` in the Agent's JSON confg file(s).
It will put this key/value into the `env-config.json` file.
When the Agent starts it will load this `env-config.json` file and use this to set a global variable via the new `SetSDKLogLevel()` function.
Anytime the Agent instantiates an `aws.Config` object it supplies `SDKLogLevel()` and `SDKLogger{}`.
Note:
- It is not useful to change `env-confg.json` after the Agent has started. The CloudWatch and CloudWatchLogs output plugins initialize their respective `aws.Config` objects once, and don't do it again.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
- Added new unit tests and modified existing test(s).
- Manual testing on Windows server 2019:
    - Set `"aws_sdk_log_level"` in Agent's JSON config (`C:\ProgramData\Amazon\AmazonCloudWatchAgent\Configs\file_config.json`)
    - Run ` & 'C:\Program Files\Amazon\AmazonCloudWatchAgent\amazon-cloudwatch-agent-ctl.ps1' -a start` from a PowerShell prompt.
    - Verify translation occured in `C:\ProgramData\Amazon\AmazonCloudWatchAgent\env-config.json`.
    - Verify messages in `C:\ProgramData\Amazon\AmazonCloudWatchAgent\Logs\amazon-cloudwatch-agent.log`.
 
```
{
	"agent": {
		"debug": true,
		"aws_sdk_log_level": "LogDebug"
	}
.
.
.
```


Sample log file with `"aws_sdk_log_level": "LogDebug"`:
```
2021/12/27 21:05:43 I! I! Detected the instance is EC2
2021/12/27 21:05:43 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json ...
/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json does not exist or cannot read. Skipping it.
2021/12/27 21:05:43 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/default ...
Valid Json input schema.
I! Detecting run_as_user...
No csm configuration found.
No log configuration found.
Configuration validation first phase succeeded
 
2021/12/27 21:05:43 I! Config has been translated into TOML /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml 
2021/12/27 21:05:43 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json ...
2021/12/27 21:05:43 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/default ...
2021/12/27 21:05:43 I! Detected runAsUser: cwagent
2021/12/27 21:05:43 I! Changing ownership of [/opt/aws/amazon-cloudwatch-agent/logs /opt/aws/amazon-cloudwatch-agent/etc /opt/aws/amazon-cloudwatch-agent/var] to 994:991
2021/12/27 21:05:43 I! Set HOME: /home/cwagent
2021-12-27T21:05:43Z I! Starting AmazonCloudWatchAgent 1.247349.0-42-g85a3de9b-dirty
2021-12-27T21:05:43Z I! AWS SDK log level, LogDebug
2021-12-27T21:05:43Z I! Loaded inputs: disk mem
2021-12-27T21:05:43Z I! Loaded aggregators: 
2021-12-27T21:05:43Z I! Loaded processors: ec2tagger
2021-12-27T21:05:43Z I! Loaded outputs: cloudwatch
2021-12-27T21:05:43Z I! Tags enabled: host=ip-172-31-5-239.ec2.internal
2021-12-27T21:05:43Z I! [agent] Config: Interval:1m0s, Quiet:false, Hostname:"ip-172-31-5-239.ec2.internal", Flush Interval:1s
2021-12-27T21:05:43Z D! [agent] Initializing plugins
2021-12-27T21:05:43Z I! [logagent] starting
2021-12-27T21:05:43Z D! Successfully created credential sessions
2021-12-27T21:05:43Z I! [processors.ec2tagger] ec2tagger: EC2 tagger has started initialization.
2021-12-27T21:05:43Z D! [agent] Connecting outputs
2021-12-27T21:05:43Z D! [agent] Attempting connection to [outputs.cloudwatch]
2021-12-27T21:05:43Z D! Successfully created credential sessions
2021-12-27T21:05:43Z I! cloudwatch: get unique roll up list []
2021-12-27T21:05:43Z D! [agent] Successfully connected to outputs.cloudwatch
2021-12-27T21:05:43Z D! [agent] Starting service inputs
2021-12-27T21:05:43Z I! cloudwatch: publish with ForceFlushInterval: 1m0s, Publish Jitter: 20s
2021-12-27T21:05:43Z D! DEBUG: Request ec2metadata/GetToken Details:
---[ REQUEST POST-SIGN ]-----------------------------
PUT /latest/api/token HTTP/1.1
Host: 169.254.169.254
User-Agent: aws-sdk-go/1.30.15 (go1.17; linux; amd64)
Content-Length: 0
X-Aws-Ec2-Metadata-Token-Ttl-Seconds: 21600
Accept-Encoding: gzip


-----------------------------------------------------
```
